lib/r2d2/util.rb

```diff
@@ -29,12 +29,12 @@ def generate_shared_secret(private_key, ephemeral_public_key)
       private_key.dh_compute_key(point)
     end
 
-    def derive_hkdf_keys(ephemeral_public_key, shared_secret, info)
+    def derive_hkdf_keys(ephemeral_public_key, shared_secret, info, key_length_bytes = 16)
       key_material = Base64.decode64(ephemeral_public_key) + shared_secret
-      hkdf_bytes = hkdf(key_material, info)
+      hkdf_bytes = hkdf(key_material, info, key_length_bytes * 2)
       {
-        symmetric_encryption_key: hkdf_bytes[0..15],
-        mac_key: hkdf_bytes[16..32]
+        symmetric_encryption_key: hkdf_bytes[0..(key_length_bytes - 1)],
+        mac_key: hkdf_bytes[key_length_bytes..(key_length_bytes * 2)]
```

Comment:
This max index is fine from a Ruby language perspective, but looks like an out-of-bounds read from a security review perspective. I suggest either adding a comment noting that indexing a string past it maximum index is non-problematic, or just use `key_length_bytes  * 2) - 1` so that it looks correct.